source/mqtt3_to_mqtt5_adapter.c

```diff
@@ -260,8 +269,451 @@ static int s_aws_mqtt5_adapter_perform_safe_callback(
+    config->port = connect_task->port;
+    config->socket_options = connect_task->socket_options;
+
+    if (config->tls_options_ptr) {
+        aws_tls_connection_options_clean_up(&config->tls_options);
+        config->tls_options_ptr = NULL;
+    }
+
+    if (connect_task->tls_options_ptr) {
+        aws_tls_connection_options_copy(&config->tls_options, connect_task->tls_options_ptr);
+        config->tls_options_ptr = &config->tls_options;
+    }
+
+    aws_byte_buf_clean_up(&adapter->client->negotiated_settings.client_id_storage);
+    aws_byte_buf_init_copy_from_cursor(
+        &adapter->client->negotiated_settings.client_id_storage,
+        adapter->allocator,
+        aws_byte_cursor_from_buf(&connect_task->client_id));
+
+    config->connect->storage_view.keep_alive_interval_seconds = connect_task->keep_alive_time_secs;
+    config->ping_timeout_ms = connect_task->ping_timeout_ms;
+    config->ack_timeout_seconds = aws_max_u64(
+        1,
+        aws_timestamp_convert(
+            connect_task->protocol_operation_timeout_ms, AWS_TIMESTAMP_MILLIS, AWS_TIMESTAMP_SECS, NULL));
+    if (connect_task->clean_session) {
+        config->session_behavior = AWS_MQTT5_CSBT_CLEAN;
+        config->connect->storage_view.session_expiry_interval_seconds = NULL;
+    } else {
+        config->session_behavior = AWS_MQTT5_CSBT_REJOIN_ALWAYS;
+        /* This is a judgement call to translate session expiry to the maximum possible allowed by AWS IoT Core */
+        config->connect->session_expiry_interval_seconds = 7 * 24 * 60 * 60;
+        config->connect->storage_view.session_expiry_interval_seconds =
+            &config->connect->session_expiry_interval_seconds;
+    }
+}
+
+static int s_aws_mqtt3_to_mqtt5_adapter_safe_connect_handler(
+    struct aws_mqtt_client_connection_5_impl *adapter,
+    void *context) {
+    struct aws_mqtt_adapter_connect_task *connect_task = context;
+
+    if (adapter->synced_data.terminated) {
+        return AWS_OP_SUCCESS;
+    }
+
+    if (adapter->adapter_state != AWS_MQTT_AS_STAY_DISCONNECTED) {
+        if (connect_task->on_connection_complete) {
+            (*connect_task->on_connection_complete)(
+                &adapter->base,
+                AWS_ERROR_MQTT_ALREADY_CONNECTED,
+                0,
+                false,
+                connect_task->on_connection_complete_user_data);
+        }
+        return AWS_OP_SUCCESS;
+    }
+
+    if (adapter->on_disconnect) {
+        (*adapter->on_disconnect)(&adapter->base, adapter->on_disconnect_user_data);
+
+        adapter->on_disconnect = NULL;
+        adapter->on_disconnect_user_data = NULL;
+    }
+
+    adapter->adapter_state = AWS_MQTT_AS_FIRST_CONNECT;
+
+    /* Update mqtt5 config */
+    s_aws_mqtt3_to_mqtt5_adapter_update_config_on_connect(adapter, connect_task);
+
+    aws_mqtt5_client_reset_connection(adapter->client);
+
+    aws_mqtt5_client_change_desired_state(adapter->client, AWS_MCS_CONNECTED, NULL);
```

Comment:
Mentioned above, I might be missing something but here we are changing the stored config options, I think without validating that they're good, and then directly changing the state to initiate a connect.

Reply:
Added some additional validation